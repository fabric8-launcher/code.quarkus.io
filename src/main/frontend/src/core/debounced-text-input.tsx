import { TextInput, TextInputProps } from '@patternfly/react-core';
import React, { useEffect, useState } from 'react';
import { useDebounce } from './use-debounce';

export interface DebouncedTextInputProps extends TextInputProps {
  delay?: number;
}

export function DebouncedTextInput(props: DebouncedTextInputProps) {
  const { value, onChange, ...rest } = props;
  const [localValue, setLocalValue] = useState(value);
  const [prevValue, setPrevValue] = useState<string | undefined>(undefined);
  const onChangeWithPrev = onChange ? (newVal: string, e: React.ChangeEvent<any>) => {
    setPrevValue(newVal);
    onChange(newVal, e);
  } : undefined;
  const debouncedOnChange = useDebounce(onChangeWithPrev, props.delay || 200);
  const onChangeWithLocal = debouncedOnChange ? (newVal: string, e: React.ChangeEvent<any>) => {
    setLocalValue(newVal);
    debouncedOnChange(newVal, e);
  }: undefined;
  useEffect(() => {
    if (value !== prevValue) {
      setLocalValue(value);
    }
    // eslint-disable-next-line
  }, [value]);
  return (
    <TextInput
      {...rest as any}
      onChange={onChangeWithLocal}
      value={onChange ? localValue: value}
    />
  );
}
