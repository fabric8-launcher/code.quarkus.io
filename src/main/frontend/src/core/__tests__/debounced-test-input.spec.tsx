import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { DebouncedTextInput } from '../debounced-text-input';
import { act } from 'react-dom/test-utils';

jest.useFakeTimers();

afterEach(() => {
  cleanup();
});

describe('<DebouncedTextInput />', () => {
  it('renders the ExtendedTextInput correctly', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    expect(comp.asFragment()).toMatchSnapshot();
    expect(comp.getByLabelText('debounced-input').getAttribute('value')).toBe('toto');
    expect(onChange).toBeCalledTimes(0);
  });

  it('Call onChange without delay when called once', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value' } });
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).lastCalledWith('some value', expect.anything());
  });

  it('Call onChange to be called after a delay for the second immediate change', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value' } });
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some new value' } });
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).lastCalledWith('some value', expect.anything());
    act(() => { jest.advanceTimersByTime(200); });
    expect(onChange).toBeCalledTimes(2);
    expect(onChange).lastCalledWith('some new value', expect.anything());
  });

  it('Call onChange to be called after a delay for the second change', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value' } });
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).lastCalledWith('some value', expect.anything());
    act(() => { jest.advanceTimersByTime(50); });
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some new value' } });
    expect(onChange).toBeCalledTimes(1);
    act(() => { jest.advanceTimersByTime(200); });
    expect(onChange).toBeCalledTimes(2);
    expect(onChange).lastCalledWith('some new value', expect.anything());
  });

  it('Call onChange only with the latest value', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value' } });
    act(() => { jest.advanceTimersByTime(50); });
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value 2' } });
    act(() => { jest.advanceTimersByTime(50); });
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value 3' } });
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).lastCalledWith('some value', expect.anything());
    act(() => { jest.advanceTimersByTime(200); });
    expect(onChange).toBeCalledTimes(2);
    expect(onChange).lastCalledWith('some value 3', expect.anything());
  });

  it('Uses the provided value on rerender', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    act(() => {
      comp.rerender(<DebouncedTextInput value="new-value" onChange={onChange} aria-label="debounced-input"/>);
    });
    expect(comp.getByLabelText('debounced-input').getAttribute('value')).toBe('new-value');
  });

  it('Uses the provided value on rerender after a change', () => {
    const onChange = jest.fn();
    const comp = render(<DebouncedTextInput value="toto" onChange={onChange} aria-label="debounced-input"/>);
    fireEvent.change(comp.getByLabelText('debounced-input'), { target: { value: 'some value' } });
    act(() => {
      comp.rerender(<DebouncedTextInput value="new-value" onChange={onChange} aria-label="debounced-input"/>);
    });
    expect(comp.getByLabelText('debounced-input').getAttribute('value')).toBe('new-value');
  });

});
