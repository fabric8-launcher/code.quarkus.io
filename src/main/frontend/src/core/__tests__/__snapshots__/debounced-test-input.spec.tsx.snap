// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DebouncedTextInput /> renders the ExtendedTextInput correctly 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    aria-label="debounced-input"
    class="pf-c-form-control"
    type="text"
    value="toto"
  />
</DocumentFragment>
`;
